lib/linked_list.rb

```diff
@@ -12,152 +11,316 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
-    end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_first(value)
-      raise NotImplementedError
-    end
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
 
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
-    def search(value)
-      raise NotImplementedError
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def add_first(value)
+    if self.length != 0
+      new_node = Node.new(value)
+      current = @head
+      @head = new_node
+      @head.next = current
+    else
+      new_node = Node.new(value)
+      @head = new_node
     end
+  end
 
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_max
-      raise NotImplementedError
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  # Time Complexity: O(n) where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def search(value)
+    if self.length == 0
+      return false
     end
 
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_min
-      raise NotImplementedError
+    current = @head
+    (self.length - 1).times do

```

Comment:
Since `.length` is a method and it traverses the list, it would be good to save that result in a variable rather than calling the method multiple times.